### PR TITLE
Implement quantization job

### LINKED ISSUE #5554: Implement model quantization job
Scope includes:
- Server-side implementation of the model quantization job
- API changes to support the job (e.g. endpoint to submit the quantization job)

### PR DESCRIPTION
**Pull request by A-Artemis**
### Summary

Implements the Quantization job.

Introduces new `JobType.QUANTIZE`
The job type requires a `max_calibration_subset_size` and an optional `max_drop`.
If `max_drop` is included, then PTQ is done with accuracy aware optimization. Otherwise the default optimization is used.

### Checklist

- [x] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: application/backend/app/api/schemas/jobs/quantization.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    model_id: UUID = Field(..., description="ID of the model revision to quantize")
+    max_calibration_subset_size: int = Field(
+        100, description="Maximum number of samples from training set used for calibration"
+    )
+    max_drop: float | None = Field(
+        None,
+        description=(
+            "Maximum allowed accuracy drop. If provided, uses nncf.quantize_with_accuracy_control(); "
+            "if omitted, uses nncf.quantize()"
```
Comment (leoll2) [inline application/backend/app/api/schemas/jobs/quantization.py]:
```suggestion
            "Maximum allowed accuracy drop, expressed as a fraction in range [0-1] where e.g. 0.03 means 3%"
```
Let's not mention implementation details in the API documentation

File: library/src/otx/backend/openvino/models/base.py [hunk 2]
```diff
@@ -238,18 +237,115 @@ def check_if_quantized(model: openvino.Model) -> bool:
...
+        quantization_dataset = nncf.Dataset(train_dataset, self.transform_fn)
+
+        if ptq_config.get("max_drop") is not None:
+            validation_dataset = nncf.Dataset(data_module.val_dataloader(), self.transform_fn)
+            validation_fn = self._create_validation_fn(data_module)
+            compressed_model = nncf.quantize_with_accuracy_control(
+                ov_model,
+                quantization_dataset,
+                validation_dataset,
+                validation_fn,
```
Comment (leoll2) [inline library/src/otx/backend/openvino/models/base.py]: I suggest to pass these arguments by keyword; in case of breaking changes in the NNCF API, it will be easier to catch them.

File: library/src/otx/backend/openvino/engine.py [hunk 3]
```diff
@@ -350,9 +353,11 @@ def optimize(
             subset="train",
         )
 
-        ptq_config = {}
+        ptq_config: dict[str, int | float] = {}
         if max_data_subset_size is not None:
             ptq_config["subset_size"] = max_data_subset_size
+        if max_drop is not None:
+            ptq_config["max_drop"] = max_drop
```
Comment (leoll2) [inline library/src/otx/backend/openvino/engine.py:363]: I suggest to log the `ptq_config` at debug level

File: application/backend/app/api/schemas/jobs/quantization.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+class QuantizationRequestParams(BaseModel):
+    """Request parameters for the quantization job."""
+
+    model_id: UUID = Field(..., description="ID of the model revision to quantize")
+    max_calibration_subset_size: int = Field(
+        100, description="Maximum number of samples from training set used for calibration"
+    )
+    max_drop: float | None = Field(
+        None,
```
Comment (leoll2) [inline application/backend/app/api/schemas/jobs/quantization.py:23]:
```suggestion
    max_drop: float | None = Field(
        None,
        ge=0,
        le=1,
```

File: application/backend/app/execution/quantization/otx_quantizer.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+            raise FileNotFoundError(f"OpenVINO model files not found at {ov_model_xml_path}")
+
+        # Check model is not already quantized (no INT8 variant exists)
+        existing_int8 = self._get_int8_variant(model)
+        if existing_int8 is not None:
+            raise ValueError(f"Model {params.model_id} already has a quantized (INT8) variant")
+
+        logger.info(
+            "Model {} validated for quantization (architecture={})", params.model_id, model.architecture
+        )  # pyrefly: ignore[missing-attribute]
```
Comment (leoll2) [inline application/backend/app/execution/quantization/otx_quantizer.py]:
```suggestion
        logger.info(
            "Model {} validated for quantization (architecture={})", model.id, model.architecture
        )
```

File: application/backend/app/execution/quantization/otx_quantizer.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+            self._dataset_revision_service.set_db_session(db)
+            dm_dataset = self._dataset_revision_service.load_revision(
+                project_id=params.project_id,
+                dataset_revision_id=dataset_revision_id,
+            )
+
+        # Extract subsets
+        dm_training_dataset = dm_dataset.filter_by_subset(Subset.TRAINING)
+        dm_validation_dataset = dm_dataset.filter_by_subset(Subset.VALIDATION)
+        dm_testing_dataset = dm_dataset.filter_by_subset(Subset.TESTING)
```
Comment (leoll2) [inline application/backend/app/execution/quantization/otx_quantizer.py:141]: As far as I understand from [NNCF docs](https://docs.openvino.ai/2026/openvino-workflow/model-optimization-guide/quantizing-models-post-training/quantizing-with-accuracy-control.html#prepare-calibration-and-validation-datasets), only our validation subset can be used for calibration/validation purposes during the optimization; the training subset is already seen by the model, while the testing subset is used for testing.

Or is the OTX engine responsible for selecting the validation dataset during optimization?

File: application/backend/app/execution/quantization/otx_quantizer.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+
+        with self._db_session_factory() as db:
+            self._project_service.set_db_session(db)
+            project = self._project_service.get_project_by_id(params.project_id)
+        task = project.task
+
+        datamodule = self.prepare_calibration_dataset(params=params, model=model)
+        ov_engine = self.initialize_engine(params=params, model=model, datamodule=datamodule)
+        try:
+            quantized_model_path = self.run_quantization(
```
Comment (leoll2) [inline application/backend/app/execution/quantization/otx_quantizer.py:287]: @kprokofi the OTX quantization interface doesn't seem very friendly; it requires to initialize a data module and engine with many parameters (see `prepare_calibration_dataset` in this file), most of which are probably not needed for quantization.

File: application/backend/app/execution/quantization/otx_quantizer.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+                return OTXTaskType.MULTI_LABEL_CLS
+            case TaskType.DETECTION:
+                return OTXTaskType.DETECTION
+            case TaskType.INSTANCE_SEGMENTATION:
+                return OTXTaskType.INSTANCE_SEGMENTATION
+            case _:
+                raise ValueError(f"Unsupported task type: {task.task_type}")
+
+    @classmethod
+    def __get_metric_by_task(cls, task: Task) -> MetricCallable:
```
Comment (leoll2) [inline application/backend/app/execution/quantization/otx_quantizer.py]: Some of these util functions are duplicated in the training job. I suggest to move them to `execution/common/otx_converters.py` or similar

File: application/backend/app/execution/quantization/otx_quantizer.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+            raise FileNotFoundError(f"Model {params.model_id} does not have an OpenVINO FP16 variant")
+
+        ov_model_xml_path = self._get_variant_model_xml_path(params.project_id, params.model_id, openvino_variant.id)
+        if not ov_model_xml_path.exists():
+            raise FileNotFoundError(f"OpenVINO model files not found at {ov_model_xml_path}")
+
+        # Check model is not already quantized (no INT8 variant exists)
+        existing_int8 = self._get_int8_variant(model)
+        if existing_int8 is not None:
+            raise ValueError(f"Model {params.model_id} already has a quantized (INT8) variant")
```
Comment (leoll2) [inline application/backend/app/execution/quantization/otx_quantizer.py]: This check should be removed: even if the model already has a quantized INT8 variant, the user should be allowed to submit another quantization job with different parameters.

File: application/backend/app/execution/quantization/otx_quantizer.py (hunk 6 above)
Comment (A-Artemis) [reply application/backend/app/execution/quantization/otx_quantizer.py:141]: Yes, the OTX engine is responsible for selecting the correct datasets. You can see that in `library/src/otx/backend/openvino/models/base.py` in the `optimize()` method.

File: application/ui/src/features/models/model-listing/current-model-training/training-model-row.component.tsx [hunk 10]
```diff
@@ -95,7 +95,8 @@ export const TrainingModelRow = ({
 }: TrainingModelRowProps) => {
     const modelId = 'model' in job.metadata ? job.metadata.model?.id : undefined;
     const { data: trainingModel } = useGetModel(modelId);
-    const modelArchitectureId = 'model' in job.metadata && job.metadata.model?.architecture;
+    const modelArchitectureId =
+        'model' in job.metadata && 'architecture' in job.metadata.model && job.metadata.model.architecture;
```
Comment (dwesolow) [inline application/ui/src/features/models/model-listing/current-model-training/training-model-row.component.tsx:99]: np. we should we `isTrainingJob` to handle this

**Review by dwesolow — APPROVED** [review]

**Review by camiloHimura — APPROVED** [review]

**Review by leoll2 — APPROVED** [review]

File: application/backend/app/execution/quantization/otx_quantizer.py (hunk 7 above)
Comment (kprokofi) [reply application/backend/app/execution/quantization/otx_quantizer.py:287]: I know it well. OTXDataModule should be rewritten in the future.

Comment (kprokofi) [reply application/backend/app/execution/quantization/otx_quantizer.py:287]: prepare_calibration_dataset looks bad in both for training and optimization. Let's create a task with low priority to track this problem

File: library/src/otx/backend/openvino/models/base.py [hunk 11]
```diff
@@ -238,18 +237,115 @@ def check_if_quantized(model: openvino.Model) -> bool:
...
+
+        Args:
+            data_module (OTXDataModule): Data module providing the validation dataloader.
+
+        Returns:
+            Callable: A function ``(compiled_model, validation_dataset) -> (float, None)``
+            where the float is the primary accuracy metric value.
+        """
+
+        def _infer_compiled_model(
```
Comment (kprokofi) [inline library/src/otx/backend/openvino/models/base.py:280]: I am not very happy to see 2 different validation pipeline TBH. We have one (engine.test) with modelAPI on the Engine level and here with compiled model on the base OV model level.
doesn't nncf support modelAPI? 
We need to think how to refactor this after release

**Review by kprokofi — APPROVED** [review]
Generally LGTM. 
However, I heard from nncf engineer that Accuracy Aware Quantization may search days for optimal solution in the worst case.
We should mention it, that user doesn't set this parameter recklessly without consideration.
We need to document that properly
cc @A-Artemis
